### PR TITLE
 [php-fpm] add option to read status from unix domain socket

### PR DESCRIPTION
**Pull request by lufia**
I added `-socket` option.
This allows to read status from Unix domain socket or TCP without HTTPd.

Standalone PHP-FPM talks FastCGI protocol, so this p-r adds an external package.

### CODE REVIEW

File: mackerel-plugin-php-fpm/testdata/tests.bash [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+run_plugin()
+{
+	sleep 1
+	go run ../main.go "$@"
+	sleep 1
+}
+
+d=$(mktemp -d -t mackerel-plugin-php-fpm)
+trap 'rm -rf $d; exit 1' 1 2 3 15
```
Comment (itchyny) [inline mackerel-plugin-php-fpm/testdata/tests.bash]: How about using EXIT trap?

File: mackerel-plugin-php-fpm/testdata/tests.bash [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
Comment (itchyny) [inline mackerel-plugin-php-fpm/testdata/tests.bash]: Isn't it more common to using .sh extension is it?

File: mackerel-plugin-php-fpm/lib/php-fpm.go [hunk 3]
```diff
@@ -17,6 +21,91 @@ type PhpFpmPlugin struct {
...
+	}
+	switch u.Scheme {
+	case "tcp":
+		p.Network = "tcp"
+		p.Address = u.Host
+	case "unix":
+		p.Network = "unix"
+		p.Address = u.Path
+	default:
+		panic("unknown scheme: " + u.Scheme)
```
Comment (itchyny) [inline mackerel-plugin-php-fpm/lib/php-fpm.go]: I think returning error is better than using panic.

File: mackerel-plugin-php-fpm/lib/php-fpm.go [hunk 4]
```diff
@@ -150,13 +247,16 @@ func Do() {
 	optLabelPrefix := flag.String("metric-label-prefix", "PHP-FPM", "Metric label prefix")
 	optTimeout := flag.Uint("timeout", 5, "Timeout")
 	optTempfile := flag.String("tempfile", "", "Temp file name")
+	var socketFlag SocketFlag
+	flag.Var(&socketFlag, "socket", "Unix domain socket `path or URL`")
```
[the remark this answers is not in the document]
Comment (lufia) [reply mackerel-plugin-php-fpm/lib/php-fpm.go]: It is intented. `SocketFlag` implements `flag.Value` interface.
When `-socket` option is passed, then `flag.Parse` calls `SocketFlag.Set` method.
Therefore if user passes invalid argument, `flag.Parse` will print usage text.

```console
$ mackerel-plugin-php-fpm -socket '://'
invalid value "://" for flag -socket: parse ://: missing protocol scheme
Usage of mackerel-plugin-php-fpm:
  -metric-key-prefix string
    	Metric key prefix (default "php-fpm")
  -metric-label-prefix string
    	Metric label prefix (default "PHP-FPM")
...
```

[the remark this answers is not in the document]
Comment (itchyny) [reply mackerel-plugin-php-fpm/lib/php-fpm.go]: Okay.

File: mackerel-plugin-php-fpm/lib/php-fpm_test.go [hunk 5]
```diff
@@ -48,3 +48,96 @@ func TestGetStatus(t *testing.T) {
...
+			Name:    "parse unix:// scheme",
+			URL:     "unix:///path/to",
+			Network: "unix",
+			Address: "/path/to",
+		},
+		{
+			Name:    "parse tcp:// scheme",
+			URL:     "tcp://localhost",
+			Network: "tcp",
+			Address: "localhost", // should we set default port?
```
Comment (itchyny) [inline mackerel-plugin-php-fpm/lib/php-fpm_test.go]: TCP with no port should be a configuration error? I'm not sure...

File: mackerel-plugin-php-fpm/lib/php-fpm.go (hunk 3 above)
Comment (lufia) [reply mackerel-plugin-php-fpm/lib/php-fpm.go]: fixed.

File: mackerel-plugin-php-fpm/lib/php-fpm_test.go (hunk 5 above)
Comment (lufia) [reply mackerel-plugin-php-fpm/lib/php-fpm_test.go]: I deleted this comment.
PHP-FPM is listen on TCP/9000 by default.

> --with-fpm-port[=PORT]       Set the tcp port number to listen for cgi requests [9000]

File: mackerel-plugin-php-fpm/testdata/tests.bash (hunk 1 above)
Comment (lufia) [reply mackerel-plugin-php-fpm/testdata/tests.bash]: fixed.

File: mackerel-plugin-php-fpm/lib/php-fpm_test.go (hunk 5 above)
Comment (itchyny) [reply mackerel-plugin-php-fpm/lib/php-fpm_test.go]: Nice.

**Comment by itchyny** [comment]
LGTM!

**Pull request merged**